core/src/main/java/org/jdbi/v3/GeneratedKeys.java

```diff
@@ -74,4 +77,9 @@
             throw new ResultSetException("Exception thrown while attempting to traverse the result set", e, context);
         }
     }
+
+    @Override
+    public <ContainerType> ContainerType collect(Class<ContainerType> containerType) {
+        return stream().collect(containerFactoryRegistry.createCollectorFor(containerType));
+    }
```

Comment:
@brianm @stevenschlansker @arteam This comment speaks to the broader discussion surrounding `Class<T>` vs `Type` parameters.

We don't get any compile-time guarantees when using `Type`. `Type` is just a marker interface, so there are no generic parameters or hints that the compiler could use to make type safety guarantees. It is a black box at compile time.

For this reason I don't think it makes sense to use `Type` in fluent APIs like e.g. Query, as in @arteam's suggestion above. The best we can offer is a method with a `Class<T>` parameter, alongside a fully generic `Whatever<T>` method e.g. `Query.map(ResultSetMapper<T>)` vs `Query.mapTo(Class<T>)`.

SQL Objects are a different story though. We have full type information available when inspecting the interface, so we can pass e.g. `getGenericReturnType()` to factories.

For this reason I propose that only factory classes (ArgumentFactory, ResultSetMapperFactory, ResultColumnMapperFactory, and CollectorFactory) be changed to use `Type` parameters.